linkerd/app/outbound/src/http/logical/router.rs

```diff
@@ -85,23 +87,108 @@ where
         // TODO(ver) use a different key type that is structured instead of
         // simply a name.
         let mut backends = HashMap::with_capacity(profile.targets.len().max(1));
-        for t in profile.targets.iter() {
+        Self::mk_backends(&mut backends, &self.new_backend, &profile.targets, &target);
+
+        // Create a stack that can distribute requests to the backends.
+        let distribute = Self::mk_distribute(&backends, &profile.targets, &target);
+        let routes = Self::mk_routes(distribute, &self.route_layer, &profile.http_routes, &target);
+
+        let (tx, rx) = watch::channel(Shared {
+            matches: profile.http_routes.clone(),
+            routes,
+        });
+
+        drop(profile);
+
+        let new_backend = self.new_backend.clone();
+        let route_layer = self.route_layer.clone();
+        let span = tracing::debug_span!("router", target = %Param::<profiles::LogicalAddr>::param(&target));
+        tokio::spawn(
+            async move {
+                while profiles.changed().await.is_ok() {
+                    let profile = profiles.borrow_and_update();
+                    // Update `backends`.
+                    // TODO(eliza): can we avoid doing this if the backends haven't
+                    // changed at all (e.g., handle updates that only change the
+                    // routes?)
+                    tracing::debug!(backends = profile.targets.len(), "Updating backends");
+                    backends.clear();
+                    Self::mk_backends(&mut backends, &new_backend, &profile.targets, &target);
```

Comment:
We need to preserve backends that are unchanged, otherwise we'll end up dropping all balancer state on updates.

Reply:
ahh, yup, you're right. will do!